### DISCUSSION TITLE
Optional Inject/Copy By Match

### DISCUSSION THREAD

**Discussion by shiribailem** [espanso/espanso#1690] [Feature requests and Ideas]
This should be a relatively simple feature given that there's already code to automatically pick one or the other per match.

Basic gist would be to add something like a "method" value to a match that takes "inject" or "copy" and overrides the input method used for that match.

I ask this because I've got a single match that needs inject, but auto picks copy. So instead I have to set all my matches to use inject for that one to work.

**Comment by smeech** [comment]
I agree - it would be useful if the `backend:` `Auto`|`Inject`|`Clipboard` could work when included in the expansion code.
See [Options](https://espanso.org/docs/configuration/options/#options-reference).
Testing seems to suggest it doesn't, sadly.

As a temporary solution, increasing the figure in `clipboard_threshold` might work. I think this sets the point at which `Inject` turns off and `Clipboard` turns on as a `backend`.

Alternatively, if your match is used in one app, and you're not in Wayland, use an [app-specific configuration](https://espanso.org/docs/configuration/app-specific-configurations/) and set the `backend` there to `Inject` or `Clipboard`, depending on which works best.

**Comment by shiribailem** [comment]
Unfortunately clipboard_threshold isn't any use for me here because the *one* match I have that requires Inject is my longest one. So to force inject on it I've got to put all of them on inject, which is why I made this request.

**Reply by smeech** [reply]
Sorry - missed your comment in my last edit, above. Would an app-specific configuration help?

**Reply by shiribailem** [reply]
It would not, these are almost all in the same app

**Reply by smeech** [reply]
Can you try `force_mode: keys` in your problematic expansion, please?

**Comment by smeech** [comment]
This is possibly solved.

Inspection of the [code](https://github.com/espanso/espanso/blob/655ebb1c8be578a57de4840ef35ce7626fc8b9e3/espanso-config/src/matches/group/loader/yaml/mod.rs#L182) reveals an additional (currently undocumented) expansion option: `force_mode: keys|clipboard`, which forces a particular `replace:` to use inject or clipboard respectively.

I would welcome some testing by others, or @federico-terzi might comment.

Without an easy way of telling which method is being used, I have resorted to an expansion:
```
  - trigger: :long
    replace: |
      xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    force_mode: clipboard
 ```
 which I use in a Thunderbird compose window, which tends to lose keystrokes if Inject (`keys`) is used, and which is visibly slower.

**Reply by AucaCoyan** [reply]
Smeech, `force_clipboard: true` is not documented to date, right? the closes point I can get is the [Injection mechanism chapter](https://espanso.org/docs/matches/basics/#injection-mechanism)

**Reply by smeech** [reply]
I consider it deprecated in favour of `force_mode: clipboard`, which does the same job. Consequently I didn't mention it in the docs. I'll change the above example to reduce confusion for anyone who comes across it.

**Reply by AucaCoyan** [reply]
Then I think it's a good idea to deprecate the force_mode, so we clean the code. I'll file an issue!

**Reply by smeech** [reply]
Deprecate `force_clipboard:`, not `force_mode:`.

**Reply by AucaCoyan** [reply]
Oh, right! I thought it right, but I've written it reversed 😆 . Issue is #1939

**Comment by shiribailem** [comment]
On first attempt this looks like it solves it exactly!

Guess the fix instead is for someone to update the documentation lol

**Comment by smeech** [comment]
Good to hear. I agree!

I'll leave this for a couple of days in case Federico wants to comment. 

The code is marked “TODO” and “test” but is two years old and a valuable feature, so I hope it's left in, although `force_mode: inject` (rather than `keys`) and scrapping of `force_clipboard: true` in favour of `force_mode: clipboard` would be more intuitive. Even more so would be the ability to use `backend: Auto|Inject|Clipboard` within expansions instead, as was my first thought.
